### PR TITLE
Migration/textfont

### PR DESCRIPTION
**Pull request by rvilarl**
Let us finish with fonts
@0xfe I can add the removal of legacy fonts to this PR

### CODE REVIEW

**Comment by rvilarl** [comment]
Help with `textWidthCacheInstance` type appreciated!

**Comment by AaronDavidNewman** [comment]
Do you mean which type should we use?  It is a 2-level hash, whose value is will be a number (float).  Does that help?  Sorry my typescript knowledge is shallow.

**Comment by rvilarl** [comment]
> Do you mean which type should we use? It is a 2-level hash, whose value is will be a number (float). Does that help? Sorry my typescript knowledge is shallow.

Thanks yes this helps and the two levels are indexed by string?

**Comment by AaronDavidNewman** [comment]
Yes, correct.

**Comment by rvilarl** [comment]
@AaronDavidNewman  any resolved, thank you!

File: src/textfont.ts [hunk 1]
```diff
@@ -9,33 +9,93 @@
...
+  protected maxSizeGlyph: string;
+  protected weight: string;
+  protected style: string;
+  protected fontCacheKey: string = '';
+
+  protected static registryInstance: TextFontRegistry[];
+  protected static textWidthCacheInstance?: Record<string, Record<string, number>>;
+
+  protected size: number;
+  protected attrs: TextFontAttributes;
```
Comment (0xfe) [inline src/textfont.ts:76]: Let's inline this.

File: src/textfont.ts [hunk 2]
```diff
@@ -9,33 +9,93 @@
...
   // Getter of an array of available fonts.  Applications may register their
   // own fonts and the metrics for those fonts will be available to the
   // application.
-  static get fontRegistry() {
+  static get fontRegistry(): TextFontRegistry[] {
     if (!TextFont.registryInstance) {
       TextFont.registryInstance = [];
       TextFont.registryInstance.push({
-        name: 'Roboto Slab',
+        name: 'RobotoSlab',
```
Comment (0xfe) [inline src/textfont.ts:98]: Is this intentional? (The space is missing.)

File: src/textfont.ts [hunk 3]
```diff
@@ -196,59 +256,55 @@ export class TextFont {
...
-    }
-    this.weight = typeof this.weight === 'undefined' ? '' : this.weight;
-    this.style = typeof this.style === 'undefined' ? '' : this.style;
     this.updateCacheKey();
   }
   // Create a hash with the current font data, so we can cache computed widths
-  updateCacheKey() {
-    this.fontCacheKey = this.family + '-' + this.size + '-' + this.weight + '-' + this.style;
+  updateCacheKey(): void {
+    this.fontCacheKey = this.family + '-' + this.size.toString() + '-' + this.weight + '-' + this.style;
```
Comment (0xfe) [inline src/textfont.ts:284]: Better to use a template string here.

File: src/textfont.ts [hunk 4]
```diff
@@ -9,33 +9,93 @@
 import { Vex } from './vex';
 import { PetalumaScriptTextMetrics } from './fonts/petalumascript_textmetrics';
 import { RobotoSlabTextMetrics } from './fonts/robotoslab_textmetrics';
+export interface TextFontMetrics {
+  advanceWidth: number;
+  ha: number;
+}
+
+export interface TextFontGlyph {
```
Comment (0xfe) [inline src/textfont.ts:17]: This looks just like the other glyph metrics. Can we share?

File: src/textfont.ts (hunk 2 above)
Comment (AaronDavidNewman) [reply src/textfont.ts:98]: Hi, the space is in the font name, probably a mistake:

`<link href="https://fonts.googleapis.com/css2?family=Roboto+Slab:wght@500&display=swap" rel="stylesheet">`

File: src/textfont.ts (hunk 4 above)
Comment (rvilarl) [reply src/textfont.ts:17]: Yes, with minor changes in `FontGlyph` done.

File: src/textfont.ts (hunk 1 above)
Comment (rvilarl) [reply src/textfont.ts:76]: done

File: src/textfont.ts (hunk 2 above)
Comment (rvilarl) [reply src/textfont.ts:98]: Not really. Reversed.

File: src/textfont.ts (hunk 3 above)
Comment (rvilarl) [reply src/textfont.ts:284]: Agree

**Comment by rvilarl** [comment]
Ready

**Comment by 0xfe** [comment]
Looks good!

**Pull request merged**